test/e2e/aws_util_test.go

```diff
@@ -0,0 +1,312 @@
+			// Check if the EIP has the specified tag key and value
+			hasTag := false
+			for _, tag := range address.Tags {
+				if *tag.Key == tagKeyEIP && *tag.Value == tagValueEIP {
+					hasTag = true
+					break
+				}
+			}
+			if hasTag {
+				// Check if the EIP is associated with a network interface in the specific VPC
+				if address.NetworkInterfaceId != nil {
+					describeNetworkInterfacesInput := &ec2.DescribeNetworkInterfacesInput{
+						NetworkInterfaceIds: []*string{address.NetworkInterfaceId},
+					}
+					describeNetworkInterfacesOutput, err := svc.DescribeNetworkInterfaces(describeNetworkInterfacesInput)
+					if err != nil {
+						t.Fatalf("failed to describe network interfaces, %v", err)
+					}
+
+					for _, networkInterface := range describeNetworkInterfacesOutput.NetworkInterfaces {
+						if networkInterface.VpcId != nil && *networkInterface.VpcId == vpcID {
+							unassociatedEIPs = append(unassociatedEIPs, struct {
+								PublicIp     string
+								AllocationId string
+							}{
+								PublicIp:     aws.StringValue(address.PublicIp),
+								AllocationId: aws.StringValue(address.AllocationId),
+							})
+						}
+					}
+				} else {
+					// If there's no associated network interface, assume it's an unassociated EIP
+					unassociatedEIPs = append(unassociatedEIPs, struct {
+						PublicIp     string
+						AllocationId string
+					}{
+						PublicIp:     aws.StringValue(address.PublicIp),
+						AllocationId: aws.StringValue(address.AllocationId),
+					})
+				}
+			}
+		} else {
+			// Check if the EIP has the specified tag key and value
+			hasTag := false
+			for _, tag := range address.Tags {
+				if *tag.Key == tagKeyEIP && *tag.Value == tagValueEIP {
+					hasTag = true
+					break
+				}
+			}
+			if hasTag {
+				// Check if the EIP is associated with a network interface in the specific VPC
+				if address.NetworkInterfaceId != nil {
+					describeNetworkInterfacesInput := &ec2.DescribeNetworkInterfacesInput{
+						NetworkInterfaceIds: []*string{address.NetworkInterfaceId},
+					}
+					describeNetworkInterfacesOutput, err := svc.DescribeNetworkInterfaces(describeNetworkInterfacesInput)
+					if err != nil {
+						t.Fatalf("failed to describe network interfaces, %v", err)
+					}
+
+					for _, networkInterface := range describeNetworkInterfacesOutput.NetworkInterfaces {
+						if networkInterface.VpcId != nil && *networkInterface.VpcId == vpcID {
+							associatedEIPs = append(associatedEIPs, struct {
+								PublicIp     string
+								AllocationId string
+							}{
+								PublicIp:     aws.StringValue(address.PublicIp),
+								AllocationId: aws.StringValue(address.AllocationId),
+							})
+						}
+					}
+				} else {
+					associatedEIPs = append(associatedEIPs, struct {
+						PublicIp     string
+						AllocationId string
+					}{
+						PublicIp:     aws.StringValue(address.PublicIp),
+						AllocationId: aws.StringValue(address.AllocationId),
+					})
+				}
+			}
+		}
+	}
```

Comment:
I'm a little confused here. There's logic that has been duplicated, it seems too complex, and the code is a bit hard to read.

1. If an EIP, that we created in this E2E, is associated, then why do we need to do further checks to see if it  associated to a network interface in our VPC? I feel like that's fair to assume that it is in our VPC. We created it, we own it.
2. If for some reason an EIP, that we created, was associated with another VPC, well that would be odd, but this logic will just ignore it? I think that's problem.
3. I could be wrong, but I don't think a EIP can *not* be associated, and have a `NetworkInterfaceId`.
4. Once an if statement gets too long, it gets unreadable in my opinion.

Is there any reason this logic couldn't be reduced to:
```
	for _, address := range describeAddressesOutput.Addresses {
		// Check if the EIP has the specified tag key and value
		hasTag := false
		for _, tag := range address.Tags {
			if *tag.Key == tagKeyEIP && *tag.Value == tagValueEIP {
				hasTag = true
				break
			}
		}
		if hasTag {
			eipAllocation := struct {
				PublicIp     string
				AllocationId string
			}{
				PublicIp:     aws.StringValue(address.PublicIp),
				AllocationId: aws.StringValue(address.AllocationId),
			})
			if address.AssociationId == nil {
			 	unassociatedEIPs = append(unassociatedEIPs, eipAllocation)
			} else {
				associatedEIPs = append(associatedEIPs, eipAllocation)
			}
		}
	}

```

Reply:
i will try this

Reply:
done thanks!